lib/iris/coords.py

```diff
@@ -2775,6 +2779,366 @@ def xml_element(self, doc):
+        self._metadata_manager = metadata_manager_factory(ConnectivityMetadata)
+
+        self._validate_arg_vs_list("start_index", start_index, [0, 1])
+        # indices array will be 2-dimensional, so must be either 0 or 1.
+        self._validate_arg_vs_list("element_dim", element_dim, [0, 1])
+        # TODO: find a better 'common' location to store valid cf_roles.
+        valid_cf_roles = [
+            "edge_node_connectivity",
+            "face_node_connectivity",
+            "face_edge_connectivity",
+            "face_face_connectivity",
+            "edge_face_connectivity",
+            "boundary_node_connectivity",
+            "volume_node_connectivity",
+            "volume_edge_connectivity",
+            "volume_face_connectivity",
+            "volume_volume_connectivity",
+        ]
+        self._validate_arg_vs_list("cf_role", cf_role, valid_cf_roles)
+
+        self._metadata_manager.start_index = start_index
+        self._metadata_manager.element_dim = element_dim
+        self._metadata_manager.cf_role = cf_role
+
+        super().__init__(
+            values=indices,
+            standard_name=None,
+            long_name=long_name,
+            var_name=var_name,
+            units="1",
+            attributes=attributes,
+        )
+
+    @staticmethod
+    def _validate_arg_vs_list(arg_name, arg, list):
+        if arg not in list:
+            error_msg = (
+                f"Invalid {arg_name} . Got: {arg} . Must be one of: "
+                f"{list} ."
+            )
+            raise ValueError(error_msg)
+
+    @staticmethod
+    def _lazy_element_lengths(array, element_dim):
+        element_mask_counts = da.sum(
+            da.ma.getmaskarray(array), axis=element_dim
+        )
+        max_element_size = array.shape[element_dim]
+        return max_element_size - element_mask_counts
+
+    @property
+    def _values(self):
+        # Overridden just to allow .setter override.
+        return super()._values
+
+    @_values.setter
+    def _values(self, values):
+        self._validate_indices(values)
+        # The recommended way of using the setter in super().
+        super(Connectivity, self.__class__)._values.fset(self, values)
+
+    @property
+    def cf_role(self):
+        """
+        The category of topological relationship that this connectivity
+        describes.
+        **Read-only** - validity of :attr:`indices` is dependent on
+        :attr:`cf_role`. A new :class:`Connectivity` must therefore be defined
+        if a different :attr:`cf_role` is needed.
+
+        """
+        return self._metadata_manager.cf_role
+
+    @property
+    def cf_role_element(self):
+        """
+        Derived from the connectivity's :attr:`cf_role` - the first part, e.g.
+        ``face`` in ``face_node_connectivity``. Refers to the elements
+        described by the structure of the connectivity's :attr:`indices` array.
+
+        """
+        return self._metadata_manager.cf_role.split("_")[0]
```

Comment:
If the `cf_role` metadata is `read-only`, then this only needs to be done once in the `__init__` and cached.

So this could be reduced to something like:
```python
return self._location
```
... if you buy into the previous comment on `cf_role_element`

Reply:
Clever. I also like the preservation of this as a property by using a private variable to store the info.

Reply:
See https://github.com/SciTools/iris/pull/3968#discussion_r570120871 for the agreed solution.

Reply:
6139e563acdd0805e72a4a3e2ffa4391c3e07337